pinot-core/src/test/java/org/apache/pinot/core/query/reduce/StreamingReduceServiceTest.java

```diff
@@ -0,0 +1,114 @@
+    when(mockedResponse.next()).thenThrow(innerException);
+    final ExecutorService threadPoolService = Executors.newFixedThreadPool(1);
+    final ServerRoutingInstance routingInstance =
+        new ServerRoutingInstance("localhost", 9527, TableType.OFFLINE, false);
+    // supposedly we can use TestNG's annotation like @Test(expectedExceptions = { IOException.class }) to verify
+    // here we hope to verify deeper to make sure the thrown exception is nested inside the exception
+    assertTrue(verifyException(() -> {
+          StreamingReduceService.processIterativeServerResponse(mock(StreamingReducer.class),
+              threadPoolService,
+              ImmutableMap.of(routingInstance, mockedResponse),
+              1000,
+              mock(BaseReduceService.ExecutionStatsAggregator.class));
+          return null;
+        }, (cause) -> cause.getMessage().contains(exceptionMessage))
+    );
+  }
+
+  @Test
+  public void testExecutionTimeout()
+      throws Exception {
+    // simulate a thread timeout in gRPC call and verify that the thread can transfer the exception
+    Iterator<Server.ServerResponse> mockedResponse = (Iterator<Server.ServerResponse>) mock(Iterator.class);
+    when(mockedResponse.hasNext()).thenReturn(true);
+    when(mockedResponse.next()).then(new Answer<Void>() {
+      @Override
+      public Void answer(InvocationOnMock invocationOnMock)
+          throws Throwable {
+        Thread.sleep(1000);
+        return null;
+      }
+    });
+    final ExecutorService threadPoolService = Executors.newFixedThreadPool(1);
+    final ServerRoutingInstance routingInstance =
+        new ServerRoutingInstance("localhost", 9527, TableType.OFFLINE, false);
+    // supposedly we can use TestNG's annotation like @Test(expectedExceptions = { IOException.class }) to verify
+    // here we hope to verify deeper to make sure the thrown exception is nested inside the exception
+    assertTrue(verifyException(() -> {
+          StreamingReduceService.processIterativeServerResponse(mock(StreamingReducer.class),
+              threadPoolService,
+              ImmutableMap.of(routingInstance, mockedResponse),
+              10,
+              mock(BaseReduceService.ExecutionStatsAggregator.class));
+          return null;
+        },
+        (cause) -> cause instanceof TimeoutException));
+  }
+
+  private static boolean verifyException(Callable<Void> verifyTarget, Function<Throwable, Boolean> verifyCause) {
+    boolean exceptionVerified = false;
+    if (verifyTarget == null || verifyCause == null) {
+      throw new RuntimeException("verifyException method needs two non-null lambdas");
+    }
+    try {
+      verifyTarget.call();
+    } catch (Exception ex) {
+      for (Throwable child = ex;
+          child != null && child.getCause() != child && !exceptionVerified;
+          child = child.getCause()) {
+        if (verifyCause.apply(child)) {
+          exceptionVerified = true;
+        }
```

Comment:
`exceptionVerified = verifyCause.test(child)` since the loop terminates as soon as the predicate passes

Reply:
Thanks. Shorter now